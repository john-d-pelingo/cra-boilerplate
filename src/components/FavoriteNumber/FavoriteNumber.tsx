import * as React from 'react'

interface IFavoriteNumberProps {
  max: number
  min: number
}

interface IFavoriteNumberState {
  numberValue: number
  numberEntered: boolean
}

class FavoriteNumber extends React.Component<
  IFavoriteNumberProps,
  IFavoriteNumberState
> {
  static defaultProps = { min: 1, max: 9 }
  state = { numberValue: 0, numberEntered: false }

  render() {
    const { numberValue, numberEntered } = this.state
    const { min, max } = this.props
    const isValid = !numberEntered || (numberValue >= min && numberValue <= max)
    return (
      <div>
        <label htmlFor="favorite-number">Favorite Number</label>
        <input
          id="favorite-number"
          type="number"
          value={numberValue}
          onChange={this.handleChange}
        />
        {isValid ? null : (
          <div data-testid="error-message">The number is invalid</div>
        )}
      </div>
    )
  }

  private handleChange = (event: React.FormEvent<HTMLInputElement>): void => {
    this.setState({
      numberEntered: true,
      numberValue: Number(event.currentTarget.value),
    })
  }
}

export default FavoriteNumber
