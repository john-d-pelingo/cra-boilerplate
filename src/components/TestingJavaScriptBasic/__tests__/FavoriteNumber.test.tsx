import 'jest-dom/extend-expect'

import React from 'react'
import { fireEvent, render } from 'react-testing-library'

import FavoriteNumber from '../FavoriteNumber'

describe('components/FavoriteNumber', () => {
  it('mounts', () => {
    const { container } = render(<FavoriteNumber />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it('entering an invalid value shows an error message', () => {
    const { getByTestId, getByLabelText, queryByTestId, rerender } = render(
      <FavoriteNumber />,
    )
    const input = getByLabelText(/favorite number/i)
    fireEvent.change(input, { target: { value: 10 } })

    expect(getByTestId('error-message')).toHaveTextContent(
      /the number is invalid/i,
    )

    rerender(<FavoriteNumber max={10} />)

    expect(queryByTestId('error-message')).not.toBeInTheDocument()
  })
})
