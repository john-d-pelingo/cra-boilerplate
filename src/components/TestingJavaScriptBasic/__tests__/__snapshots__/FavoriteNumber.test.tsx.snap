// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/FavoriteNumber mounts 1`] = `
<div>
  <label
    for="favorite-number"
  >
    Favorite Number
  </label>
  <input
    id="favorite-number"
    type="number"
    value="0"
  />
</div>
`;
